{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "043f66a4",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-01T23:51:12.313371Z",
     "start_time": "2025-02-01T23:51:12.311787Z"
    }
   },
   "outputs": [],
   "source": [
    "import os\n",
    "from pathlib import Path\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7525fdc4",
   "metadata": {},
   "source": [
    "# Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "008da993",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-01T23:51:14.707240Z",
     "start_time": "2025-02-01T23:51:14.702989Z"
    }
   },
   "outputs": [],
   "source": [
    "class SingleOmicsDataset:\n",
    "    def __init__(self, path: str, name: str, sep='\\t'):\n",
    "        self.path = path\n",
    "        self.name = name\n",
    "        self.sep = sep\n",
    "        self.data = None\n",
    "        \n",
    "        self._load_data()\n",
    "        \n",
    "    def _load_data(self):\n",
    "        self.data = pd.read_csv(self.path, sep=self.sep, index_col=0)\n",
    "        self.data = self.data.transpose()\n",
    "        self.data.index.names = ['sample']\n",
    "        self.data.columns.names = ['feature']\n",
    "    \n",
    "    def remove_missing_labels(self, label_indices):\n",
    "        self.data = self.data[self.data.index.isin(label_indices)]\n",
    "        self.data = self.data.sort_index(axis=0)\n",
    "\n",
    "    def clean_missing_values(self):\n",
    "        # Keep features/columns with less than 10% missing values\n",
    "        min_non_na = int(len(self.data.index) * 0.9)\n",
    "        has_missing_values = self.data.isna().any().any()\n",
    "        self.data = self.data.dropna(axis=1, thresh=min_non_na)\n",
    "        \n",
    "        # Fill NA values with the mean of each feature/column\n",
    "        self.data = self.data.fillna(self.data.mean())\n",
    "\n",
    "        # Check for any missing values in the DataFrame\n",
    "        has_missing_values = self.data.isna().any().any()\n",
    "        if has_missing_values:\n",
    "            raise ValueError(\"The modality contains missing values. Please handle them before proceeding.\")\n",
    "\n",
    "    def normalize_data(self):\n",
    "        # apply the min-max scaling for each feature separately\n",
    "        self.data = (self.data - self.data.min()) / (self.data.max() - self.data.min())\n",
    "\n",
    "    def remove_low_variance_features(self, threshold):\n",
    "        if threshold is not None:\n",
    "            self.data = self.data.loc[:, self.data.var() >= threshold]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8d63f708",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-02-01T23:51:21.192720Z",
     "start_time": "2025-02-01T23:51:21.185464Z"
    }
   },
   "outputs": [],
   "source": [
    "class MultiomicsDataset:\n",
    "    def __init__(self, \n",
    "                 raw_data_dir,\n",
    "                 raw_label_dir,\n",
    "                 num_omics,\n",
    "                 sep='\\t',\n",
    "                 clean_missing=True,\n",
    "                 normalize=True,\n",
    "                 var_threshold=None,\n",
    "                 label_column_name=None,\n",
    "                 label_column_values=None):\n",
    "        self.raw_data_dir = raw_data_dir\n",
    "        self.raw_label_dir = raw_label_dir\n",
    "        self.label_column_name = label_column_name\n",
    "        self.label_column_values = label_column_values\n",
    "        self.sep = sep\n",
    "        self.num_omics = num_omics\n",
    "        self.clean_missing = clean_missing\n",
    "        self.normalize = normalize\n",
    "        self.var_threshold = var_threshold\n",
    "        self.data = []\n",
    "        self.label = None\n",
    "        \n",
    "        self._process()\n",
    "        \n",
    "    def _process(self):\n",
    "        if self.label_column_name is not None:\n",
    "            self.label = pd.read_csv(self.raw_label_dir, sep=self.sep, index_col=0)\n",
    "            self.label = self.label[self.label[self.label_column_name].notnull()]\n",
    "            self.label = self.label[self.label_column_name]\n",
    "            if self.label_column_values is not None:\n",
    "                self.label = self.label[self.label.isin(self.label_column_values)]\n",
    "        print(f\"\\nInput class labels:\\n{self.label.value_counts(dropna=True)}\\n\")\n",
    "        \n",
    "        self.label = self.label.sort_index(axis=0)\n",
    "\n",
    "        for path, name in self.raw_data_dir:\n",
    "            omics_data = SingleOmicsDataset(path, name, self.sep)\n",
    "            self.data.append(omics_data)\n",
    "        self.print_omics_data(\"Raw/Input omic modalities\")\n",
    "\n",
    "        # remove samples with missing labels (their labels are not included in the label file)\n",
    "        for omics_data in self.data:\n",
    "            omics_data.remove_missing_labels(self.label.index)\n",
    "        self.print_omics_data(\"After missing label removal\")\n",
    "\n",
    "        # check whether each index in label is present in at least one of the omic dataset\n",
    "        is_available, labels_to_remove = self._check_label_indices_availability()\n",
    "        self.label = self.label.drop(labels_to_remove)\n",
    "        print(f\"Are all labels available in omic modalities? {is_available}\\n\")\n",
    "        \n",
    "        # clean missing values\n",
    "        if self.clean_missing:\n",
    "            for omics_data in self.data:\n",
    "                omics_data.clean_missing_values()\n",
    "            self.print_omics_data(\"After missing value removal\")\n",
    "\n",
    "        # normalize features\n",
    "        if self.normalize:\n",
    "            for omics_data in self.data:\n",
    "                omics_data.normalize_data()\n",
    "            \n",
    "        # remove low variance features\n",
    "        if self.var_threshold is not None:\n",
    "            for omics_idx, omics_data in enumerate(self.data):\n",
    "                omics_data.remove_low_variance_features(self.var_threshold[omics_idx])\n",
    "            self.print_omics_data(\"After low-variance feature removal\")\n",
    "    \n",
    "    def _check_label_indices_availability(self):\n",
    "        # Create an empty set to store indices from all omics data\n",
    "        combined_omics_indices = set()\n",
    "        \n",
    "        # Union indices from each omic dataset\n",
    "        for omics_data in self.data:\n",
    "            combined_omics_indices.update(omics_data.data.index)\n",
    "\n",
    "        not_combined_omics_count = 0\n",
    "        labels_to_remove = []\n",
    "\n",
    "        # Check if each label index is in the combined omics indices\n",
    "        for label_idx in self.label.index:\n",
    "            if label_idx not in combined_omics_indices:\n",
    "                not_combined_omics_count += 1\n",
    "                labels_to_remove.append(label_idx)\n",
    "\n",
    "        print(f\"Number of labels not in combined omics indices: {not_combined_omics_count}\")\n",
    "\n",
    "        return not_combined_omics_count == 0, labels_to_remove\n",
    "\n",
    "    def map_labels(self):\n",
    "        unique_labels = self.label.unique()\n",
    "        label_to_number = {label: i for i, label in enumerate(unique_labels)}\n",
    "        self.label = self.label.map(label_to_number)\n",
    "        self.label = self.label.rename(\"Class\")\n",
    "        print(f\"Mapped class labels: {label_to_number}\")\n",
    "        print(f\"\\nMapped class labels:\\n{self.label.value_counts(dropna=False)}\\n\")\n",
    "\n",
    "    def save_result(self, save_label=True):\n",
    "        print(\"Saving the processed data...\")\n",
    "        if save_label:\n",
    "            processed_label_path = self.raw_label_dir.replace(\"raw_data\", \"processed_data\") + \".csv\"\n",
    "            print(f\" - label path: {processed_label_path}, label shape:{self.label.shape}\")\n",
    "            self.label.to_csv(processed_label_path, index=True)\n",
    "\n",
    "        for omics_idx, omics_data in enumerate(self.data):\n",
    "            processed_path = self.raw_data_dir[omics_idx][0].replace(\"raw_data\", \"processed_data\") + \".csv\"\n",
    "            print(f\" - {self.raw_data_dir[omics_idx][1]} path: {processed_path}, {self.raw_data_dir[omics_idx][1]} shape: {omics_data.data.shape}\")\n",
    "            omics_data.data.to_csv(processed_path, index=True)\n",
    "\n",
    "    def print_omics_data(self, title):\n",
    "        print(f\"\\nOmic modality shape ({title}):\")\n",
    "        for omics_data in self.data:\n",
    "            print(f\" - {omics_data.name} shape: {omics_data.data.shape}\")\n",
    "        print()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d023290",
   "metadata": {},
   "source": [
    "# Breast Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "cc05e268",
   "metadata": {},
   "outputs": [],
   "source": [
    "base_raw_dir = os.path.join(\"..\", \"data\", \"raw_data\", \"BRCA\")\n",
    "base_processed_dir = os.path.join(\"..\", \"data\", \"processed_data\", \"BRCA\")\n",
    "\n",
    "raw_data_dir = [\n",
    "    (os.path.join(base_raw_dir, \"DNA\"), \"DNA\"), # DNA methylation HumanMethylation27\n",
    "    (os.path.join(base_raw_dir, \"mRNA\"), \"mRNA\"), # Gene expression RNAseq\n",
    "    (os.path.join(base_raw_dir, \"miRNA\"), \"miRNA\"), # miRNA mature strand expression RNAseq\n",
    " ]\n",
    "raw_label_dir = os.path.join(base_raw_dir, \"ClinicalMatrix\")\n",
    "\n",
    "if not os.path.exists(base_processed_dir):\n",
    "    os.makedirs(base_processed_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "bef182de",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Input class labels:\n",
      "PAM50Call_RNAseq\n",
      "LumA      434\n",
      "LumB      194\n",
      "Basal     142\n",
      "Normal    119\n",
      "Her2       67\n",
      "Name: count, dtype: int64\n",
      "\n",
      "\n",
      "Omic modality shape (Raw/Input omic modalities):\n",
      " - DNA shape: (345, 27578)\n",
      " - mRNA shape: (1218, 20530)\n",
      " - miRNA shape: (832, 2238)\n",
      "\n",
      "\n",
      "Omic modality shape (After missing label removal):\n",
      " - DNA shape: (328, 27578)\n",
      " - mRNA shape: (956, 20530)\n",
      " - miRNA shape: (584, 2238)\n",
      "\n",
      "Number of labels not in combined omics indices: 0\n",
      "Are all labels available in omic modalities? True\n",
      "\n",
      "\n",
      "Omic modality shape (After missing value removal):\n",
      " - DNA shape: (328, 24956)\n",
      " - mRNA shape: (956, 20530)\n",
      " - miRNA shape: (584, 436)\n",
      "\n",
      "\n",
      "Omic modality shape (After low-variance feature removal):\n",
      " - DNA shape: (328, 6860)\n",
      " - mRNA shape: (956, 2608)\n",
      " - miRNA shape: (584, 436)\n",
      "\n",
      "Mapped class labels: {'Normal': 0, 'LumA': 1, 'LumB': 2, 'Basal': 3, 'Her2': 4}\n",
      "\n",
      "Mapped class labels:\n",
      "Class\n",
      "1    434\n",
      "2    194\n",
      "3    142\n",
      "0    119\n",
      "4     67\n",
      "Name: count, dtype: int64\n",
      "\n",
      "Saving the processed data...\n",
      " - label path: ../data/processed_data/BRCA/ClinicalMatrix.csv, label shape:(956,)\n",
      " - DNA path: ../data/processed_data/BRCA/DNA.csv, DNA shape: (328, 6860)\n",
      " - mRNA path: ../data/processed_data/BRCA/mRNA.csv, mRNA shape: (956, 2608)\n",
      " - miRNA path: ../data/processed_data/BRCA/miRNA.csv, miRNA shape: (584, 436)\n"
     ]
    }
   ],
   "source": [
    "multiomics_data = MultiomicsDataset(raw_data_dir, \n",
    "                                    raw_label_dir,\n",
    "                                    label_column_name='PAM50Call_RNAseq',\n",
    "                                    clean_missing=True,\n",
    "                                    normalize=True,\n",
    "                                    num_omics=3,\n",
    "                                    var_threshold=[0.04, 0.03, None])\n",
    "multiomics_data.map_labels()\n",
    "multiomics_data.save_result()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a875b190",
   "metadata": {},
   "source": [
    "# Bladder Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "a7b62000",
   "metadata": {},
   "outputs": [],
   "source": [
    "base_raw_dir = os.path.join(\"..\", \"data\", \"raw_data\", \"BLCA\")\n",
    "base_processed_dir = os.path.join(\"..\", \"data\", \"processed_data\", \"BLCA\")\n",
    "\n",
    "raw_data_dir = [\n",
    "    (os.path.join(base_raw_dir, \"DNA\"), \"DNA\"), # DNA methylation HumanMethylation450\n",
    "    (os.path.join(base_raw_dir, \"mRNA\"), \"mRNA\"), # Gene expression RNAseq\n",
    "    (os.path.join(base_raw_dir, \"miRNA\"), \"miRNA\"), # miRNA mature strand expression RNAseq\n",
    " ]\n",
    "raw_label_dir = os.path.join(base_raw_dir, \"ClinicalMatrix\")\n",
    "\n",
    "if not os.path.exists(base_processed_dir):\n",
    "    os.makedirs(base_processed_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "3f77dcaa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Input class labels:\n",
      "neoplasm_histologic_grade\n",
      "High Grade    412\n",
      "Low Grade      21\n",
      "Name: count, dtype: int64\n",
      "\n",
      "\n",
      "Omic modality shape (Raw/Input omic modalities):\n",
      " - DNA shape: (434, 485577)\n",
      " - mRNA shape: (426, 20530)\n",
      " - miRNA shape: (429, 2210)\n",
      "\n",
      "\n",
      "Omic modality shape (After missing label removal):\n",
      " - DNA shape: (431, 485577)\n",
      " - mRNA shape: (423, 20530)\n",
      " - miRNA shape: (426, 2210)\n",
      "\n",
      "Number of labels not in combined omics indices: 0\n",
      "Are all labels available in omic modalities? True\n",
      "\n",
      "\n",
      "Omic modality shape (After missing value removal):\n",
      " - DNA shape: (431, 395362)\n",
      " - mRNA shape: (423, 20530)\n",
      " - miRNA shape: (426, 471)\n",
      "\n",
      "\n",
      "Omic modality shape (After low-variance feature removal):\n",
      " - DNA shape: (431, 8592)\n",
      " - mRNA shape: (423, 5980)\n",
      " - miRNA shape: (426, 471)\n",
      "\n",
      "Mapped class labels: {'High Grade': 0, 'Low Grade': 1}\n",
      "\n",
      "Mapped class labels:\n",
      "Class\n",
      "0    412\n",
      "1     21\n",
      "Name: count, dtype: int64\n",
      "\n",
      "Saving the processed data...\n",
      " - label path: ../data/processed_data/BLCA/ClinicalMatrix.csv, label shape:(433,)\n",
      " - DNA path: ../data/processed_data/BLCA/DNA.csv, DNA shape: (431, 8592)\n",
      " - mRNA path: ../data/processed_data/BLCA/mRNA.csv, mRNA shape: (423, 5980)\n",
      " - miRNA path: ../data/processed_data/BLCA/miRNA.csv, miRNA shape: (426, 471)\n"
     ]
    }
   ],
   "source": [
    "multiomics_data = MultiomicsDataset(raw_data_dir, \n",
    "                                    raw_label_dir,\n",
    "                                    label_column_name='neoplasm_histologic_grade',\n",
    "                                    num_omics=3,\n",
    "                                    var_threshold=[0.08, 0.03, None])\n",
    "multiomics_data.map_labels()\n",
    "multiomics_data.save_result()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5623e133",
   "metadata": {},
   "source": [
    "# Ovarian Cancer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "027c2623",
   "metadata": {},
   "outputs": [],
   "source": [
    "base_raw_dir = os.path.join(\"..\", \"data\", \"raw_data\", \"OV\")\n",
    "base_processed_dir = os.path.join(\"..\", \"data\", \"processed_data\", \"OV\")\n",
    "\n",
    "raw_data_dir = [\n",
    "    (os.path.join(base_raw_dir, \"DNA\"), \"DNA\"), # DNA methylation HumanMethylation27\n",
    "    (os.path.join(base_raw_dir, \"mRNA\"), \"mRNA\"), # Gene expression RNAseq\n",
    "    (os.path.join(base_raw_dir, \"miRNA\"), \"miRNA\"), # miRNA mature strand expression RNAseq\n",
    " ]\n",
    "raw_label_dir = os.path.join(base_raw_dir, \"ClinicalMatrix\")\n",
    "modified_label_dir = os.path.join(base_raw_dir, \"ClinicalMatrix_modified\")\n",
    "\n",
    "if not os.path.exists(base_processed_dir):\n",
    "    os.makedirs(base_processed_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "0e41f875",
   "metadata": {},
   "outputs": [],
   "source": [
    "def discretize_class(label):\n",
    "    if (label['days_to_death'] / 365) >= 3:\n",
    "        return \"long-term\"\n",
    "    elif label['vital_status'] == 'DECEASED':\n",
    "        return \"short-term\"\n",
    "    return None\n",
    "\n",
    "label = pd.read_csv(raw_label_dir, sep='\\t', index_col=0)\n",
    "label = label[label['days_to_death'].notnull()]\n",
    "label['Class'] = label.apply(discretize_class, axis=1)\n",
    "label = label[label['Class'] != None].loc[:, ['Class']]\n",
    "label.to_csv(modified_label_dir, sep='\\t', index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "476a16fc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "Input class labels:\n",
      "Class\n",
      "short-term    184\n",
      "long-term     177\n",
      "Name: count, dtype: int64\n",
      "\n",
      "\n",
      "Omic modality shape (Raw/Input omic modalities):\n",
      " - DNA shape: (616, 27578)\n",
      " - mRNA shape: (308, 20530)\n",
      " - miRNA shape: (485, 2165)\n",
      "\n",
      "\n",
      "Omic modality shape (After missing label removal):\n",
      " - DNA shape: (356, 27578)\n",
      " - mRNA shape: (184, 20530)\n",
      " - miRNA shape: (302, 2165)\n",
      "\n",
      "Number of labels not in combined omics indices: 1\n",
      "Are all labels available in omic modalities? False\n",
      "\n",
      "\n",
      "Omic modality shape (After missing value removal):\n",
      " - DNA shape: (356, 24876)\n",
      " - mRNA shape: (184, 20530)\n",
      " - miRNA shape: (302, 448)\n",
      "\n",
      "\n",
      "Omic modality shape (After low-variance feature removal):\n",
      " - DNA shape: (356, 5167)\n",
      " - mRNA shape: (184, 11308)\n",
      " - miRNA shape: (302, 448)\n",
      "\n",
      "Mapped class labels: {'long-term': 0, 'short-term': 1}\n",
      "\n",
      "Mapped class labels:\n",
      "Class\n",
      "1    184\n",
      "0    176\n",
      "Name: count, dtype: int64\n",
      "\n",
      "Saving the processed data...\n",
      " - label path: ../data/processed_data/OV/ClinicalMatrix_modified.csv, label shape:(360,)\n",
      " - DNA path: ../data/processed_data/OV/DNA.csv, DNA shape: (356, 5167)\n",
      " - mRNA path: ../data/processed_data/OV/mRNA.csv, mRNA shape: (184, 11308)\n",
      " - miRNA path: ../data/processed_data/OV/miRNA.csv, miRNA shape: (302, 448)\n"
     ]
    }
   ],
   "source": [
    "multiomics_data = MultiomicsDataset(raw_data_dir, \n",
    "                                    modified_label_dir,\n",
    "                                    label_column_name='Class',\n",
    "                                    num_omics=3,\n",
    "                                    var_threshold=[0.04, 0.03, None])\n",
    "multiomics_data.map_labels()\n",
    "multiomics_data.save_result()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.14"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
